### ISSUE TITLE
tries to read object from the wrong cache space

### ISSUE THREAD

**Issue by acarl005** [CoursePark/NestHydrationJS#15]
Hello @dvalentiate . I've used this library for a remarkably long time without any issues. It's robust, performant, and very easy to use. I've found some unexpected behavior that might be a bug.

I have the following:
```js
const rawData = [
 {
   id: '430',
   webs__id: '23685',
   webs__rightsManagement__id: '23689',
   webs__rightsManagement__policy: 'monetize',
   webs__rightsManagement__contracts__id: '23697' },
 {
   id: '430',
   webs__id: '23686',
   webs__rightsManagement__id: '23690',
   webs__rightsManagement__policy: 'monetize',
   webs__rightsManagement__contracts__id: '23698' },
 {
   id: '430',
   webs__id: '23687',
   webs__rightsManagement__id: '23691',
   webs__rightsManagement__policy: 'monetize',
   webs__rightsManagement__contracts__id: '23699' },
 {
   id: '430',
   webs__id: '23688',
   webs__rightsManagement__id: '23692',
   webs__rightsManagement__policy: 'monetize',
   webs__rightsManagement__contracts__id: '23700' } ]

const shapeDef = {
 id: 'id',
 webs:
  { id: 'webs__id',
    rightsManagement:
     { id: 'webs__rightsManagement__id',
       policy: 'webs__rightsManagement__policy',
       contracts: { id: 'webs__rightsManagement__contracts__id' } } } }

const { nest } = require('nesthydrationjs')
const nestedData = nest(rawData, shapeDef)
```

This works fine 😀. As expected, I get the following nested data:
```js
{ id: '430',
  webs:
   { id: '23685',
     rightsManagement: { id: '23689', policy: 'monetize', contracts: [Object] } } }
```

If you look at the `shapeDef` variable, it specifies that `web` is a single object rather than a "many".  The `rawData`, however, had ids for multiple different `web`s. But it produced the expected result, which is that it kept `'23685'` and ignored the rest.

The problem arises when I do the following: 
```js
const shapeDef = {
 id: 'id',
 webs:
  { id: 'webs__id',
    rightsManagement:
     { id: 'webs__rightsManagement__id',
       policy: 'webs__rightsManagement__policy',
       // HERE i change contracts to a "many"
       contracts: [ { id: 'webs__rightsManagement__contracts__id' } ] } } }
```

If I re-run the snippet with that change, I get the following error:
```
TypeError: Cannot read property 'contracts' of undefined
    at _nest (/Users/andy/Documents/stem/alpha/node_modules/nesthydrationjs/NestHydrationJS.js:179:14)
    at NestHydrationJS.nest (/Users/andy/Documents/stem/alpha/node_modules/nesthydrationjs/NestHydrationJS.js:205:4)
    at repl:1:20
```

I debugged a little bit. It looks like when it checked the cache for `rightsManagement` with id `'23690'`, which is nested under the *second* `web`, the `container` was `undefined`, causing the error.

When I changed `web` to a "many", the error went away.

Is this the intended outcome? I would've expected that instead of an exception I'd get only `web` with id `'23685'` with all its nested contracts.

**Comment by dvalentiate** [comment]
well you certainly know how to butter me up, the compliments are much appreciated. Even more appreciated is the detailed example with sample data.

The bad news is I'm about to go on vacation for 2 weeks... I'm definitely interested in solving this though.

**Comment by dvalentiate** [comment]
I also like the the term `shape` for the definition. Variable naming, one of the hardest of programming problems.

**Comment by dvalentiate** [comment]
TL;DR
Requested behaviour isn't a feature of this library, the author finds the idea interesting, but it needs to be considered more.


ok looking at your example in more detail I understand what is going on.

I hadn't considered this library would be used for something similar to a `group by, use first` on purpose. It does work for that on leaf object properties, but that isn't a feature, just coincidental behaviour. I don't have any tests for it. I had also never considered what would happen if one had tried to get a `group by first` operation to work in the middle of the data structure.

While I can imagine it being possible to support this behaviour, it would be difficult because usually the library only knows about the current data row it is reading and tries to navigate the output data structure based on that. The `group by, use first` would require remembering past navigation paths or traversing the output data structure. The first method would have memory impacts, the other has performance impacts (but only if that code path is used) and would be complicated to code.

I should have probably output an intelligible error if the `group by, use first` situation was encountered.

I might be able to enable some `group by` functionality with some extension to the definition so that only navigation paths for those specified in the definition are remembered (cached). This would keep the memory footprint of the library down, and wouldn't be too messy. But that is still a pretty big task.

**Comment by acarl005** [comment]
@dvalentiate Having worked a lot on an open source project myself for the past year, I know some of the OSS laments. People tend to open issues when there is, well, an issue. So you hear a disproportionate amount of negative feedback. It's really nice when people take a moment to say some positives too.

Yeah I realize that the behavior when multiple IDs are encountered when the single object was specified in the definition was not documented. It made no such promises of any `group by, use first ` functionality. My expectations were formed from trials and experimentation. I appreciate that supporting this behavior will have memory impacts and you don't want to adversely affect the performance. 

For now, I'll take care to bring back only a single ID in data sets that expect to have a single record.

**Comment by dvalentiate** [comment]
@acarl005 thanks for understanding.

Oh and I haven't used your [join-monster](https://github.com/stems/join-monster) but I'm certainly thinking about it now that I've looked into it!
